### ISSUE TITLE
Support structs and enums in ABI

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#595]
Implement the following specification: https://github.com/ethereum/EIPs/issues/50

**Comment by chriseth** [comment]
This most likely also fixes https://github.com/ethereum/solidity/issues/980

**Comment by chriseth** [comment]
Just a note to myself: The encoder for tuples / structs should work as follows:
```
// nextfree is the place where we can allocate memory,
// all pointers are absolute memory positions
encodeTupleToMemory_t1t2t3...tn(from, to, nextfree) -> (next, nextfree_)
{
    encode values from "from" to "to", passing and updating "nextfree"
    store offsets relative to "to"
    return (from + 'head size', updated nextfree)
}
```

**Comment by axic** [comment]
This has been implemented as of 0.4.19.
